src/Microsoft.ComponentDetection.Detectors/rust/RustCrateDetector.cs

```diff
@@ -34,61 +38,200 @@ protected override Task OnFileFound(ProcessRequest processRequest, IDictionary<s
-                    return Task.CompletedTask;
+                    throw new FormatException($"Could not find matching package for dependency string '{dependency}'");
+                }
+
+                if (IsLocalPackage(childPackage))
+                {
+                    if (!IsLocalPackage(parentPackage))
+                    {
+                        throw new FormatException($"In package with source '{parentComponent.Id}' found non-source dependency string: '{dependency}'");
+                    }
+
+                    // This is a dependency between packages without source
+                    return;
                 }
 
-                var cargoPackages = cargoLock.package.ToHashSet();
-                RustCrateUtilities.BuildGraph(cargoPackages, cargoDependencyData.NonDevDependencies, cargoDependencyData.DevDependencies, singleFileComponentRecorder);
+                var detectedComponent = new DetectedComponent(childComponent);
+                seenAsDependency.Add(childPackage);
+
+                if (IsLocalPackage(parentPackage))
+                {
+                    // We are adding a root edge (from a local package)
+                    singleFileComponentRecorder.RegisterUsage(detectedComponent, isExplicitReferencedDependency: true);
+                }
+                else
+                {
+                    // we are adding an edge within the graph
+                    singleFileComponentRecorder.RegisterUsage(detectedComponent, isExplicitReferencedDependency: false, parentComponentId: parentComponent.Id);
+                }
             }
             catch (Exception e)
             {
-                // If something went wrong, just ignore the file
+                using var record = new RustCrateDetectorTelemetryRecord();
+
+                record.PackageInfo = $"{parentPackage.name}, {parentPackage.version}, {parentPackage.source}";
+                record.Dependencies = dependency;
+
                 Logger.LogFailedReadingFile(cargoLockFile.Location, e);
             }
+        }
+
+        private static readonly Regex DependencyFormatRegex = new Regex(
+           ////  PkgName[ Version][ (Source)]
+           @"^([^ ]+)(?: ([^ ]+))?(?: \(([^()]*)\))?$",
+           RegexOptions.Compiled);
+
+        private const int PackageNameGroup = 1;
+        private const int VersionGroup = 2;
+        private const int SourceGroup = 3;
+
+        private static bool ParseDependency(string dependency, out string packageName, out string version, out string source)
+        {
+            var match = DependencyFormatRegex.Match(dependency);
+            packageName = match.Groups[PackageNameGroup].Success ? match.Groups[PackageNameGroup].Value : null;
+            version = match.Groups[VersionGroup].Success ? match.Groups[VersionGroup].Value : null;
+            source = match.Groups[SourceGroup].Success ? match.Groups[SourceGroup].Value : null;
```

Comment:
nit: named groups[^1]

```suggestion
           @"^(?<packageName>[^ ]+)(?: (?<version>[^ ]+))?(?: \((?<source>[^()]*)\))?$",
           RegexOptions.Compiled);

        private static bool ParseDependency(string dependency, out string packageName, out string version, out string source)
        {
            var match = DependencyFormatRegex.Match(dependency);
            packageName = match.Groups["packageName"].Success ? match.Groups["packageName"].Value : null;
            version = match.Groups["version"].Success ? match.Groups["version"].Value : null;
            source = match.Groups["source"].Success ? match.Groups["source"].Value : null;
```

[^1]: https://docs.microsoft.com/en-us/dotnet/standard/base-types/grouping-constructs-in-regular-expressions#named_matched_subexpression